### ISSUE TITLE
IMobileServiceClient.LoginAsync - how it use?

### ISSUE THREAD

**Issue by FetFrumos** [Azure/azure-mobile-apps-net-client#341]
Hi, Azure Team.
I used azure mobile net client 3.1.0.0. This is my old code:

   IMobileServiceClient client;
  //for Android
  //from Microsoft.WindowsAzure.Mobile.Ext
   var user = await client.LoginAsync(Forms.Context, provider);

Now I have updated to the new version 4.0. I need authorization with Google account. This is the code I found in the documentation:

    IMobileServiceClient client;
    var token = new JObject {{"access_token", "access_token_value"}};
    user = await client.LoginAsync(provider, token);

How do I get  "access_token_value" on client Android Phone?
Please help me. A small example or reference to the documentation. 

Thank you.

**Comment by MatkovIvan** [comment]
Hi @FetFrumos!

You need to perform steps similar to [android documentation](https://docs.microsoft.com/en-us/azure/app-service-mobile/app-service-mobile-android-get-started-users).
1. [Add your app to the Allowed External Redirect URLs](https://docs.microsoft.com/en-us/azure/app-service-mobile/app-service-mobile-android-get-started-users#a-nameredirecturlaadd-your-app-to-the-allowed-external-redirect-urls)
1. Add 3rd parameter to `LoginAsync` call
```
var user = await client.LoginAsync(Forms.Context, provider, "{url_scheme_of_your_app}");
```
2. Add `RedirectUrlActivity` to `AndroidManifest.xml`
```
<activity android:name="com.microsoft.windowsazure.mobileservices.authentication.RedirectUrlActivity"
          android:launchMode="singleTop" android:noHistory="true">
    <intent-filter>
        <action android:name="android.intent.action.VIEW" />
        <category android:name="android.intent.category.DEFAULT" />
        <category android:name="android.intent.category.BROWSABLE" />
        <data android:scheme="{url_scheme_of_your_app}" android:host="easyauth.callback" />
    </intent-filter>
</activity>
```
We are in the process of updating our documentation.

**Comment by FetFrumos** [comment]
Hi, @MatkovIvan
I use 

           Microsoft.WindowsAzure.MobileServices.IMobileServiceClient client;

from assembly Microsoft.Azure.Mobile.Client 4.0.0.0.
In this type there is no function with three parameters.
Should I use a different type?

**Comment by MatkovIvan** [comment]
This is extension method for implementation of this interface - `Microsoft.WindowsAzure.MobileServices.MobileServiceClient`. Just store client as `MobileServiceClient` instead `IMobileServiceClient`.

**Comment by FetFrumos** [comment]
I found MobileServiceClient. Thank you.
{url_scheme_of_your_app} -  is it  URL of my application appended with the path, /.auth/login/google/callback?

**Comment by MatkovIvan** [comment]
Nope. Scheme is only part of [URL](https://en.wikipedia.org/wiki/URL#Syntax). Callback URL will be `{url_scheme_of_your_app}://easyauth.callback`. You can use any URL scheme you choose. It should be unique to your mobile application. This allow the authentication to system redirect back to your app once the authentication process is complete.
For example:
`http` - associated with system browser.
`myawesomeapp` - will be associated with some awesome app because `activity` with `intent-filter` related to this scheme added to `AndroidManifest.xml`.

**Comment by FetFrumos** [comment]
I am very sorry. I changed {url_scheme_of_your_app} to http:myapp7.  Now I run method LoginAsync()   and get  browser with text: "You have successfully with signed". This window hangs constantly. When I press the back button I get exception :" Authentication was cancelled by the user."
  Maybe I need to update my Azure mobile app(backend)?

**Comment by MatkovIvan** [comment]
> The scheme, consisting of a sequence of characters beginning with a letter and followed by any combination of letters, digits, plus (+), period (.), or hyphen (-). Although schemes are case-insensitive, the canonical form is lowercase and documents that specify schemes must do so with lowercase letters. It is followed by a colon (:).

Remove `http:`. It should be just `myapp7` or something like that.
You should you `try`/`catch` block to handle cancellation by the user (via back button for example).

**Comment by FetFrumos** [comment]
This is my exception (when click back button)

     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task 
     task) [0x0003e] in <3fd174ff54b146228c505f23cf75ce71>:0 
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification 
    (System.Threading.Tasks.Task task) [0x00028] in <3fd174ff54b146228c505f23cf75ce71>:0 
    at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) 
    [0x00008] in <3fd174ff54b146228c505f23cf75ce71>:0 
    at System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () [0x00000] in 
    <3fd174ff54b146228c505f23cf75ce71>:0 
     at Microsoft.WindowsAzure.MobileServices.MobileServicePKCEAuthentication+
     <LoginAsyncOverride>d__14.MoveNext () [0x0006f] in <b11f53647e2e407897b89c0e1a8d3bc1>:0 
      --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in 
       <3fd174ff54b146228c505f23cf75ce71>:0 
       at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task 
       task) [0x0003e] in <3fd174ff54b146228c505f23cf75ce71>:0 
       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification 
       (System.Threading.Tasks.Task task) [0x00028] in <3fd174ff54b146228c505f23cf75ce71>:0 
       at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) 
       [0x00008] in <3fd174ff54b146228c505f23cf75ce71>:0 
       at System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () [0x00000] in 
      <3fd174ff54b146228c505f23cf75ce71>:0 
       at Microsoft.WindowsAzure.MobileServices.MobileServiceAuthentication+
      <LoginAsync>d__25.MoveNext () [0x00068] in <b11f53647e2e407897b89c0e1a8d3bc1>:0 
      --- End of stack trace from previous location where exception was thrown ---
        at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in 
     <3fd174ff54b146228c505f23cf75ce71>:0 
      at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task 
      task) [0x0003e] in <3fd174ff54b146228c505f23cf75ce71>:0 
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification 
       (System.Threading.Tasks.Task task) [0x00028] in <3fd174ff54b146228c505f23cf75ce71>:0 
      at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) 
      [0x00008] in <3fd174ff54b146228c505f23cf75ce71>:0 
      at System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () [0x00000] in 
      <3fd174ff54b146228c505f23cf75ce71>:0 
      at XDayTimer.Droid.Service.Sync.Authentication+<LoginAsync>d__0.MoveNext () [0x00077] in 
     H:\MyPath..\MyAuth.cs:10

**Comment by MatkovIvan** [comment]
@FetFrumos, as I wrote earlier you should use `try`/`catch` block to handle cancellation by the user.
```C# 
try {
    var user = await client.LoginAsync(Forms.Context, provider, "{url_scheme_of_your_app}");
} catch (InvalidOperationException e) {
    // User cancelled authentication
}
```

**Comment by FetFrumos** [comment]
But I need  variable user. In block catch user always null.

     try
     {
         var user = await client.LoginAsync(Forms.Context, provider, "{url_scheme_of_your_app}");
      } 
       catch (InvalidOperationException e) 
      {
         // User cancelled authentication
         //user always = null
      }

**Comment by MatkovIvan** [comment]
If user canceled login, it means that he is NOT logged and you can't use user variable.

**Comment by FetFrumos** [comment]
This is my user case:
1. Click "Sign in" in app. This run :
       var user = await client.LoginAsync(...);
2. Next, the browser opens with the text "You have successfully signed in".
This window is displayed all the time.  I'm not returning to the application.
If I press the back button, I get an exception and user == null.

**Comment by MatkovIvan** [comment]
Please double check `intent-filter` in `AndroidManifest.xml`.

**Comment by FetFrumos** [comment]
It is work - thank you very much for your help.

**Comment by 12michel34** [comment]
Hi,
i tried to implement the workflow on my Xamrin.Forms app. I used a uriScheme like "abc-def" for the app and added "abc-def://easyauth.callback" to the backend as allowed callback url. If i try to login while dbuging on my android device everything seems to work fine (Login screen appears, etc) but then a browser opens with the message: "abc-def://easyauth.callback/#authorization_code=averylongstring" is not reachable ERR_UNKNOWN_URL_SCHEME.

**Comment by FetFrumos** [comment]
Hi,  @12michel34. Try uriScheme   without "-". Simply "abcdef". Check your manifest AndroidManifest.xml.  I there had an error.

**Comment by 12michel34** [comment]
Thanks, @FetFrumos . My mistake was: i had to place the RedirectUrlActivity inside the application tag of the AndroidManifest.xml. Now the android app is working and i have to find out how to make this working on UWP and iOS.... :o)

**Comment by utloggad2** [comment]
I got it working on iOS, the trick seems to be add a URL Scheme to the info.plist, and then add an overloaded method of OpenUrl where a call to the Mobile Client should made as follows 
`public override bool OpenUrl(UIApplication application, NSUrl url, string sourceApplication, NSObject annotation)
        {
            // return base.OpenUrl(application, url, sourceApplication, annotation);
            return <your_mobile_client_v4>.ResumeWithURL(url);
        }`
@12michel34 , I don't get it to work on Android. The mobile client only has 2 exposed methods for LoginAsync on Android and none of them is what I want, I need to be able to LoginAsync with some additional parameters in order to get offline access to Google for instance. How did you accomplish the LoginAsync?

**Comment by 12michel34** [comment]
@utloggad2 , My mobile client provides 6 overloads for LoginAsync(). Do you have the right version (4.0.0) installed on your android app?

**Comment by utloggad2** [comment]
Yes, I have the 4.0.0 client. I scratched my head for a while but a restart of Visual Studio fixed it. I can see all the methods now. 

Skickat från min iPhone

**Comment by danielloczi** [comment]
Hi,
I would like to use Google authentication with the v4 Mobile Client inside a UWP app. I registered the URL Scheme, the login window come up, I can login with a gmail account, and after that my UWP app continues, but not in the next code after LoginAsync(), but it calls into the OnActivated method. So the login task never finishes. Can you help me, how to solve this? Thx.

**Comment by bmourat** [comment]
Hi @danielloczi 
It is correct that your app is resuming from OnActivated method. After you call ResumeWithURL (this will finish login task) on the client object, your code after LoginAsync should execute. You can check out UWP login example here: https://github.com/dhei/azure-mobile-apps-xamarinforms-login-sample

**Comment by AndreyBespamyatnov** [comment]
Still, the same issue, after I logging I see only white screen

![image](https://user-images.githubusercontent.com/2779266/27917329-5e7674d8-6274-11e7-8b3e-77c501e973da.png)

**Comment by FlxGen** [comment]
Same here with the white screen.  I'm in iOS.  I feel like its related to the app url but everything I have matches the (half assed) docs.  Why was this change necessary?  Seems a lot more complicated then the view and provider....

**Comment by AndreyBespamyatnov** [comment]
I have reverted SDK version to 3.1,  because at 4.0 auth not working. at 4.0 version i see only white screen

**Comment by bmourat** [comment]
@AndreyBespamyatnov could you please share some details of your code. How are you trying to authenticate?

**Comment by AndreyBespamyatnov** [comment]
Hi @bmourat, 

Yes sure, that it is all works at version **3** not in **4**

0) I want to update my Azure SDK to new version **4**
1) `IMobileServiceClient` was changet to `Microsoft.WindowsAzure.MobileServices.MobileServiceClient`
2) Add `RedirectUrlActivity` to `AndroidManifest.xml` inside `application` tag
`<activity android:name="com.microsoft.windowsazure.mobileservices.authentication.RedirectUrlActivity"
          android:launchMode="singleTop" android:noHistory="true">
    <intent-filter>
        <action android:name="android.intent.action.VIEW" />
        <category android:name="android.intent.category.DEFAULT" />
        <category android:name="android.intent.category.BROWSABLE" />
        <data android:scheme="fsrapp" android:host="easyauth.callback" />
    </intent-filter>
</activity>`
3) Add fsrapp schema to my azure auth settings
![image](https://user-images.githubusercontent.com/2779266/27956871-63a81b0e-6324-11e7-893d-54f9d0c8d5c8.png)
4) run my xamarin forms application (android platform), i can login but i see white screen
![image](https://user-images.githubusercontent.com/2779266/27956901-8efcc0de-6324-11e7-9465-aef75623a2bb.png)
5) Try to debug that, and i see that i have loggined and i have current user at my context, but application wont closing(hiding) `RedirectUrlActivity` activity.

**Comment by bmourat** [comment]
@AndreyBespamyatnov one more thing, could your provide your replyUrl settings in your AAD app?

**Comment by AndreyBespamyatnov** [comment]
Sure @bmourat (Thanks for help!!!)

this is: https://fsrapp20170704022705.azurewebsites.net/.auth/login/aad/callback

**Comment by bmourat** [comment]
@AndreyBespamyatnov 
Seems like this is known [issue](https://github.com/Azure/azure-mobile-services-xamarin-auth/pull/8). It is already fixed and will be in the next release which should be soon (maybe this week or next).

**Comment by AndreyBespamyatnov** [comment]
@bmourat wow, great news, let's wait for new release.

**Comment by BioTurboNick** [comment]
Since this is about CustomTabs, maybe this is the same issue as #363?

**Comment by den2107** [comment]
@AndreyBespamyatnov  double check if you have placed the  <intent-filter> INSIDE the <application> node at the AndroidManifest.

I had the same problem tha you said here and it was because i had placed the <intent-filter> after the </application>

**Comment by EmilAlipiev** [comment]
@den2107 why cant they just write this on the official documentation i dont really understand. entire this azure login process is too complicated. i did the same process with aws cognito in 1 hour.

**Comment by Sophophilia** [comment]
@AndreyBespamyatnov  Did you ever get it working? I am still stuck where you are, sign in is successful, but it never returns to my app... I have 2 test devices. My Pixel looks like yours, White screen.
![screenshot_20180127-154321](https://user-images.githubusercontent.com/17054459/35485683-14e5c92c-0431-11e8-8396-50cb10b4d6ec.png)

 But with my old moto g this is what it looks like
![moto error azureauth](https://user-images.githubusercontent.com/17054459/35485674-ec753748-0430-11e8-807d-5313ee139374.png)

**Comment by Pinox** [comment]
@Sophophilia  ,  I face exactly the same issue as you.  Followed the docs and recommendation on this list. Android app does not redirect back to the app. Can someone plse confirm that this is indeed working?

**Comment by bmourat** [comment]
Hi @Pinox,

I've just checked with completely new test app and everything works fine. To investigate your problem could you please share the following info:
- Allowed external redirect uri registered on the server side
- Type of project (Android, Xamarin.Android, Xamarin.Forms)
- Login code
- AndroidManifest file

Thanks!

**Comment by Pinox** [comment]
Thanks @bmourat 

My Azure Redirect :     (using anonymized app name but format the same)
https://mytestapp-authentication.azurewebsites.net://easyauth.callback

Xamarin Forms App 
Running latest stable XF with links to .net standard libs version 2. XF App can build and release fine. I upgraded from the old azure mobile 3.0 libs where it used to work fine

Login code
   var user = await client.LoginAsync(MainActivity.Instance,   provider  , "https://mytestapp-authentication.azurewebsites.net");

Login seems to work fine as browser is showing the return token same as @Sophophilia in previous post  but redirect does not happen back to android app.

Android manifest
![androidmanifest](https://user-images.githubusercontent.com/5297255/35806701-1b46aa4e-0a89-11e8-918d-8a53db97db43.PNG)

Thanks for helping.

**Comment by MatkovIvan** [comment]
Hi @Pinox,

Your redirect URL looks wrong. You should use **scheme** before `://easyauth.callback`.
Example of correct redirect URL: `mytestapp-authentication://easyauth.callback`.
You can read more about URL syntax on wiki: https://en.wikipedia.org/wiki/URL#Syntax

Best,
Ivan

**Comment by Pinox** [comment]
Hi @MatkovIvan ,

Just for clarity I have used the following values as per my understanding

Azure Portal Redirect Url  : https://mytestapp-authentication.azurewebsites.net://easyauth.callback

Android manifest file Redirect: 
    <data android:scheme="mytestapp-authentication" android:host="easyauth.callback" />

Still not working on my side.

**Comment by MatkovIvan** [comment]
@Pinox Replace `https://mytestapp-authentication.azurewebsites.net://easyauth.callback` to `mytestapp-authentication://easyauth.callback`

**Comment by bmourat** [comment]
@Pinox please make sure that you use only scheme part in the login call, like this
```
var user = await client.LoginAsync(MainActivity.Instance, provider , "mytestapp-authentication");
```

**Comment by Pinox** [comment]
@bmourat @MatkovIvan   Thanks a lot, really much appreciated. It's finally working, one helluva journey to get there.  For anyone reading this a short summary.

If your authentication app on Azure is running at https://mytestapp-authentication.azurewebsites.net

Your Azure Redirect should be:          mytestapp-authentication://easyauth.callback

Your Android manifest callback should include a line :  
data android:scheme="mytestapp-authentication" android:host="easyauth.callback"

and lastly your LoginAsync should be:
var user = await client.LoginAsync(MainActivity.Instance, provider , "mytestapp-authentication");

I tested above setup on all providers and they all redirect back to Android app.

**Comment by ltaljaard** [comment]
@Pinox Hi, could you kindly confirm that you have set a value of  mytestapp-authentication://easyauth.callback under portal.azure.com -> mytestapp -> Authentication / Authorization -> ALLOWED EXTERNAL REDIRECT URLS ? I've done that but it won't save on that screen, it says "Failed to save Auth Settings for mytestapp: ... Cannot update the site '{0}' because Authentication / Authorization was configured with an invalid external URL ..."

**Comment by jerryno6** [comment]
Hi all !

I've just finished this class: [AZR120](https://university.xamarin.com/classes#authentication-with-azure-azr120).

I used material from this [Exercise_1_Complete](https://github.com/XamarinUniversity/AZR120/tree/master/Exercise%201/Completed)
and upgraded the Nuget component named AzureMobileClient to version 4.0.2.0, 
I tried to update the manifest, change the uriScheme
`return azureClient.LoginAsync(Android.App.Application.Context, provider, "mysecurepersonaldiary");`

But after all, I cannot get the app worked with version 4.x like it used to be. It stuck at the webview with state: **You have successfully signed in.**

Could you please give us the update source code of this exercise which use AzureMobileClient v4.x ?

**Comment by routsanjoo** [comment]
Hi ,
Our existing MobileApps (Xamarin Native) built on  using Azure Mobile Service Client sdk 3.0 unfortunately showing the  warning message before login screen appear, please suggest any work around.

The warning is this  "Security Warning : there are problems with the security certificate for this site"

**Comment by aldincedeno** [comment]
Hi @bmourat @MatkovIvan 

I am using the new authentication experience (no classic), but it does not allow me to add a custom external url in Allowed external redirect URLs (in the format: mytestapp-authentication://easyauth.callback)

It gives me that error message,

Thanks in advance for your comments

![AutenticacionAzure](https://user-images.githubusercontent.com/13489904/115986499-bf76e780-a5b0-11eb-91d8-6007f5a24baf.png)

**Comment by MatkovIvan** [comment]
Murat and I aren't owning this project anymore.

The initial question of this (closed) topic was solved a while ago, so I suggest you start a new issue and fill template/all details so the team can take a look and help you.
